src/unix/linux-core.c

```diff
@@ -793,27 +793,62 @@ uint64_t uv_get_total_memory(void) {
 }
 
 
-static uint64_t uv__read_cgroups_uint64(const char* cgroup, const char* param) {
-  char filename[256];
+static uint64_t uv__read_uint64(const char* filename) {
   char buf[32];  /* Large enough to hold an encoded uint64_t. */
   uint64_t rc;
 
   rc = 0;
-  snprintf(filename, sizeof(filename), "/sys/fs/cgroup/%s/%s", cgroup, param);
   if (0 == uv__slurp(filename, buf, sizeof(buf)))
-    sscanf(buf, "%" PRIu64, &rc);
+    if (1 != sscanf(buf, "%" PRIu64, &rc))
+      if (0 == strcmp(buf, "max\n"))
+        rc = ~0ull;
 
   return rc;
 }
 
 
+/* This might return 0 if there was a problem getting the memory limit from
+ * cgroups. This is OK because a return value of 0 signifies that the memory
+ * limit is unknown.
+ */
+static uint64_t uv__get_constrained_memory_fallback(void) {
+  return uv__read_uint64("/sys/fs/cgroup/memory/memory.limit_in_bytes");
+}
+
+
 uint64_t uv_get_constrained_memory(void) {
-  /*
-   * This might return 0 if there was a problem getting the memory limit from
-   * cgroups. This is OK because a return value of 0 signifies that the memory
-   * limit is unknown.
-   */
-  return uv__read_cgroups_uint64("memory", "memory.limit_in_bytes");
+  char filename[4097];
+  char buf[1024];
+  uint64_t high;
+  uint64_t max;
+  char* p;
+
+  if (uv__slurp("/proc/self/cgroup", buf, sizeof(buf)))
+    return uv__get_constrained_memory_fallback();
+
+  /* TODO(bnoordhuis) Scan in case there are multiple lines? */
+  if (memcmp(buf, "0::/", 4))
+    return uv__get_constrained_memory_fallback();
```

Comment:
This is the only part I'm not really sure about. I think there can be more lines but I suspect the first line is always the root hierarchy. Please correct me if I'm wrong.

Reply:
This should be fine for cgroupv2 detection. According to kernel docs:

> `/proc/$PID/cgroup` lists a process's cgroup membership.  If legacy
> cgroup is in use in the system, this file may contain multiple lines,
> one for each hierarchy.  The entry for cgroup v2 is always in the
> format `0::$PATH`:
>
> ```
>  # cat /proc/842/cgroup
>  ...
>  0::/test-cgroup/test-cgroup-nested
> ```